Error Handling in Microflows - create a new document (#5332)
mendix/docs: content/en/docs/refguide/modeling/application-logic/microflows-and-nanoflows/microflows/error-handling-in-microflows.md

aliases:
    - /howto/logic-business-rules/set-up-error-handling.html
    - /howto/logic-business-rules/set-up-error-handling
---

## 1 Introduction

In a [microflow](/refguide/microflows/), microflow elements ([activities](/refguide/activities/), [decisions](/refguide/decisions/), or [loops](/refguide/loop/)) can sometimes fail and produce an error. If you do nothing, Mendix will report the error with a system error message and roll back all the changes. However, you can also change this behavior and handle errors yourself.

> This document introduces different error handling options in microflows and explains how they work using simple examples. An example of how combined custom error handling works is also given in this document.

Reviewer: Link to section (also needs anchor added, below)
```suggestion
This document introduces different error handling options in microflows and explains how they work using simple examples. An example of how combined custom error handling works is also given in [An Example of Combined Error Handling](#example).
```

This document also presents information on how you can inspect errors using predefined error objects and the best practices for working with error handling in microflows.

> ## 2 Error Handling Options {#errorhandlers}

Author: Explain that this does not apply to the changes applied within the activity:
Emir Uner [5 days ago](https://paper.dropbox.com/doc/Error-Handling-in-Microflows-the-latest-version--BspuiTqNxxKdn9kaI4iWtYGNAg-jPf6V0YsUzdlApQmzLhU9?t=852442347648348478634223#:t=852442347648348478634223)
​​Nothing is rolled back​ may not be entirely correct. If some objects are created inside GenerateError​ they are actually rolled back.

Paul Moes [4 days ago](https://paper.dropbox.com/doc/Error-Handling-in-Microflows-the-latest-version--BspuiTqNxxKdn9kaI4iWtYGNAg-jPf6V0YsUzdlApQmzLhU9?t=852442347648348478634223#:t=852442347648348478634223)
​​I think this is a point that should be addressed somewhere in this doc. When you set rollback on a microflow it does not apply to what goes on inside that microflow, but only to what happens up to that point.
RG
Robbert Grootjans [4 days ago](https://paper.dropbox.com/doc/Error-Handling-in-Microflows-the-latest-version--BspuiTqNxxKdn9kaI4iWtYGNAg-jPf6V0YsUzdlApQmzLhU9?t=852442347648348478634223#:t=852442347648348478634223)
​​This is specifically about ‘Continue’ I believe, and that it is stated that ‘Nothing is rolled back’. Similar to ‘Custom without Rollback’ a savepoint is created before the call to GenerateError​ and upon error everything is moved back to the state of that savepoint. In contrast with ‘Custom without rollback’ you cannot do any handling. (@MarkvanMents )

Reviewer: New paragraph added in section 2.3, below.

There are four error handling options:

* **Rollback** (default)
* **Custom with rollback**
* **Custom without rollback**
* **Continue**

> Except for the **Rollback** option, which is the default, the other three are all custom error handling options. For **Custom with rollback** and **Custom without rollback**, you can create an additional flow from the microflow elements (activities or decisions) and mark this flow as an error handling flow. You can end this flow with an [error event](/refguide/error-event/) or an [end event](/refguide/end-event/). Hence, we introduce these two custom error handling options based on what event terminates the error handling flow.

Reviewer: ```suggestion
Except for the **Rollback** option, which is the default, the other three are all custom error handling options. For **Custom with rollback** and **Custom without rollback**, you can create an additional flow from the microflow elements (activities or decisions) from which the error is returned and mark this flow as an error handling flow. You can end this flow with an [error event](/refguide/error-event/) or an [end event](/refguide/end-event/). Hence, we introduce these two custom error handling options based on what event terminates the error handling flow.
```

{{% alert color="info" %}}

> In the error handling flow, you can include activities that update the database. Custom error handling does not affect newly created objects there, but if you want to update the status of an existing object that was created outside of this flow and that was rolled back because of custom error handling, then an error will occur.

Reviewer: ```suggestion
In the error handling flow, you can include activities that update the database. Custom error handling does not affect newly created objects there, but if you try to update the status of an existing object that was created outside of this flow and that was rolled back because of custom error handling, then an error will occur.
```

### 2.1 Error Handling  - Default

In this example, the error handling in **Microflow 2** is the default: **Rollback**.

{{< figure src="/attachments/refguide/modeling/application-logic/microflows-and-nanoflows/microflows/error-handling-in-microflows/default-roll-back.png" width="600px" >}}

When **Microflow 1** starts with a button click, a savepoint is created at the very beginning of **Microflow 1**. When an error occurs in **Microflow 2**,  the default error handling is to roll back all changes to the state of this savepoint (so all changes are reverted), the microflow is aborted, the error is logged, and a system error message is shown to the end-user.

### 2.2 Error Handling - Custom with Rollback

#### 2.2.1 Error Handling Which Ends with an Error Event

In this example, the error handling in **Microflow 2** is set to **Custom with rollback** and the error handling flow ends with an error event.

{{< figure src="/attachments/refguide/modeling/application-logic/microflows-and-nanoflows/microflows/error-handling-in-microflows/with-roll-back-error-event.png" width="600px" >}}

> When **Microflow 1** starts with a button click, a savepoint is created at the very beginning of **Microflow 1**. When an error occurs in **Microflow 2**, everything rolls back to the state of this savepoint because the error handling is set to **Custom with rollback**. Hence, changes made in ****Create Order**** and in **Create Customer** are both reverted. The error is logged using a custom error message. After that, the error event throws an error to terminate **Microflow 2**, and **Microflow 1** will terminate because there is no custom error handing there.

Reviewer: ```suggestion
When **Microflow 1** starts with a button click, a savepoint is created at the very beginning of **Microflow 1**. When an error occurs in **Microflow 2**, everything rolls back to the state of this savepoint because the error handling is set to **Custom with rollback**. Hence, changes made in **Create Order** and in **Create Customer** are both reverted. A custom error is logged using a **Log message** activity. After that, the error event throws an error to terminate **Microflow 2**, and **Microflow 1** will terminate because there is no custom error handing there.
```

#### 2.2.2 Error Handling Which Ends with an End Event

In this example, the error handling in **Microflow 2** is set to **Custom with rollback** and the error handling flow ends with an end event.

{{< figure src="/attachments/refguide/modeling/application-logic/microflows-and-nanoflows/microflows/error-handling-in-microflows/with-roll-back-end-event.png" width="600px" >}}

> When **Microflow 1** starts with a button click, a savepoint is created at the very beginning of **Microflow 1**. When an error occurs in **Microflow 2**, everything rolls back to the state of this savepoint because the error handling is set to **Custom with rollback**. Hence, changes made in **Create Order** and in **Create Customer** are both reverted. The error is logged using a custom error message.

Reviewer: ```suggestion
When **Microflow 1** starts with a button click, a savepoint is created at the very beginning of **Microflow 1**. When an error occurs in **Microflow 2**, everything rolls back to the state of this savepoint because the error handling is set to **Custom with rollback**. Hence, changes made in **Create Order** and in **Create Customer** are both reverted. A custom error is logged using a **Log message** activity.
```

> ### 2.3 Error Handling - Custom without Rollback

Reviewer: Response to discussion above: https://github.com/mendix/docs/pull/5332/files#r1029573045

```suggestion
### 2.3 Error Handling - Custom without Rollback

Setting "Custom without Rollback" will not stop data changes within the activity which failed from being rolled back. For example, if a [Call REST Service](/refguide/call-rest-action/) returns an error, any data the REST call retrieved will not be kept.
```

#### 2.3.1 Error Handling Which Ends with an Error Event

In this example, the error handling in **Microflow 2** is set to **Custom without rollback** and the error handling flow ends with an error event.

{{< figure src="/attachments/refguide/modeling/application-logic/microflows-and-nanoflows/microflows/error-handling-in-microflows/without-roll-back-error-event.png" width="600px" >}}

> When **Microflow 1** starts with a button click, a savepoint is created at the very beginning of **Microflow 1**. Another savepoint is created right before **GenerateError** because the error handling is set to **Custom without rollback**. When an error occurs in **Microflow 2**, changes made in **Create Order** are at this moment still kept because of the savepoint right before **GenerateError**. The error is logged using a custom error message. After that, the error event throws an error to terminate **Microflow 2** and rolls back everything to the state of savepoint which is at the very beginning of **Microflow 1**. Hence, changes made in **Create Customer** and in **Create Order** are both reverted.

Reviewer: ```suggestion
When **Microflow 1** starts with a button click, a savepoint is created at the very beginning of **Microflow 1**. Another savepoint is created right before **GenerateError** because the error handling is set to **Custom without rollback**. When an error occurs in **Microflow 2**, changes made in **Create Order** are at this moment still kept because of the savepoint right before **GenerateError**. A custom error is logged using a **Log message** activity. After that, the error event throws an error to terminate **Microflow 2** and rolls back everything to the state of savepoint 1 which is at the very beginning of **Microflow 1**. Hence, changes made in **Create Customer** and in **Create Order** are both reverted.
```

{{% alert color="info" %}}
Using custom error handling without rollback and ending the error handling flow with an error event here makes it possible to create a custom log message with details about the **NewOrder** (if you used error handling with rollback, this information would not be available). However, after logging this message, everything will still be rolled back.  
{{% /alert %}}

#### 2.3.2 Error Handling Which Ends with an End Event

In this example, the error handling in **Microflow 2** is set to **Custom without rollback** and the error handling flow ends with an end event.

{{< figure src="/attachments/refguide/modeling/application-logic/microflows-and-nanoflows/microflows/error-handling-in-microflows/without-roll-back-end-event.png" width="600px" >}}

> When **Microflow 1** starts with a button click, a savepoint is created at the very beginning of **Microflow 1**. Another savepoint is created right before **GenerateError** because the error handling is set to **Custom without rollback**. When an error occurs in **Microflow 2**, changes made in **Create Order** are saved because of the savepoint right before **GenerateError**. The error is logged using a custom error message. **Microflow 1** does not receive the error and continues to its end. Changes made in **Create Customer** are saved.

Reviewer 2: There are still a few places where 'saved' is used instead of 'kept'.

Reviewer: ```suggestion
When **Microflow 1** starts with a button click, a savepoint is created at the very beginning of **Microflow 1**. Another savepoint is created right before **GenerateError** because the error handling is set to **Custom without rollback**. When an error occurs in **Microflow 2**, changes made in **Create Order** are kept because of the savepoint right before **GenerateError**. A custom error is logged using a **Log message** activity. **Microflow 1** does not receive the error and continues to its end. Changes made in **Create Customer** are kept.
```

### 2.4 Error Handling – Continue

The **Continue** option can only be set on a microflow call or on a loop.

In this example, the error handling in **Microflow 2** is set to **Continue**.

{{< figure src="/attachments/refguide/modeling/application-logic/microflows-and-nanoflows/microflows/error-handling-in-microflows/continue.png" width="600px" >}}

> When **Microflow 1** starts with a button click, a savepoint is created at the very beginning of **Microflow 1**. Another savepoint is created right before **GenerateError** because the error handling is set to **Continue**. When the error in **Microflow 2** occurs, the microflow continues as if nothing happened. Changes made in **Create Customer** and **Create Order** are both kept. No error message is logged and no message is displayed to the end-user.

Reviewer: Response to https://github.com/mendix/docs/pull/5332/files#r1029573045

```suggestion
When **Microflow 1** starts with a button click, a savepoint is created at the very beginning of **Microflow 1**. Another savepoint is created right before **GenerateError** because the error handling is set to **Continue**. When the error in **Microflow 2** occurs, the microflow continues as if nothing happened. Changes made in **Create Customer** and **Create Order** are both kept. No error message is logged and no message is displayed to the end-user.

Data changes within the activity which returns an error (for example, within an iteration of a [loop](/refguide/loop/)) will always be rolled back. It is only data changes outside the activity returning the error which will be kept.
```

{{% alert color="warning" %}}
You should be very careful with using the **Continue** option since it can make it very difficult to understand when something goes wrong. Where possible, try to use **Custom without rollback** instead and log the error message.
{{% /alert %}}

> ## 3 An Example of Combined Error Handling

Reviewer: Anchor for link added above.
```suggestion
## 3 An Example of Combined Error Handling {#example}
```

In this example, the error handling on the **GenerateError** activity in **Microflow 2** and on the call of **Microflow 2** are both set to **Custom without rollback**. The error handling flow in **Microflow 2** ends with an error event. 

{{< figure src="/attachments/refguide/modeling/application-logic/microflows-and-nanoflows/microflows/error-handling-in-microflows/combination-without-roll-back-error-event.png" width="600px" >}}

There are three savepoints. When **Microflow 1** starts with a button click, a savepoint is created at the very beginning of **Microflow 1**. Another savepoint is created right before the call of **Microflow 2**, and the third savepoint is created right before **GenerateError** - this is because the error handling on the **GenerateError** activity in **Microflow 2** and on the call of **Microflow 2** are both set to **Custom without rollback**.

> When an error occurs in **Microflow 2**, changes made in **Create Order** are at this moment saved because of the savepoint right before **GenerateError**. The error is logged using a custom error message. After that, the error event throws an error to terminate **Microflow 2** and rolls back everything inside **Microflow 2** to the state of **savepoint 2**, (at this moment, what was saved in **Create Order** is now rolled back). The custom error handling in **Microflow 1** that applies to the call of **Microflow 2** causes the error handling flow in **Microflow 1** to be followed and terminates **Microflow 1** with an end event. Changes made in **Create Customer** are hence kept.

Reviewer: ```suggestion
When an error occurs in **Microflow 2**, changes made in **Create Order** are at this moment saved because of the savepoint right before **GenerateError**. A custom error is logged using a **Log message** activity. After that, the error event throws an error to terminate **Microflow 2** and rolls back everything inside **Microflow 2** to the state of **savepoint 2**, (at this moment, what was saved in **Create Order** is now rolled back). The custom error handling in **Microflow 1** that applies to the call of **Microflow 2** causes the error handling flow in **Microflow 1** to be followed and terminates **Microflow 1** with an end event. Changes made in **Create Customer** are hence kept.
```

> ## 4 Inspecting Errors

Author: Paul Moes [4 days ago](https://paper.dropbox.com/doc/Error-Handling-in-Microflows-the-latest-version--BspuiTqNxxKdn9kaI4iWtYGNAg-jPf6V0YsUzdlApQmzLhU9?t=908419779813674140346866#:t=908419779813674140346866)
​​maybe we can mention errors when calling a rest service? In that situation they need to inspect the latestHttpResponse variable [Call REST Service | Mendix Documentation](https://docs.mendix.com/refguide/call-rest-action/#latesthttpresponse) @MarkvanMents

Reviewer: I'll rewrite this section completely and link to the commit below.

Reviewer: Added in https://github.com/mendix/docs/pull/5332/commits/6fc62ac012d6bf15676afc65d6e7be8cd90ca4de

> In microflows that apply entity access, it is not possible to inspect the attributes of error objects for security reasons. You can pass the error object to a sub-microflow that does not apply entity access and inspect the attributes there.

Author: Paul Moes [4 days ago](https://paper.dropbox.com/doc/Error-Handling-in-Microflows-the-latest-version--BspuiTqNxxKdn9kaI4iWtYGNAg-jPf6V0YsUzdlApQmzLhU9?t=963933578239613488863277#:t=963933578239613488863277)
​​I’m not sure how this works with entity access enabled, but with entity access disabled you can’t pass the error message to a sub-microflow. That would result in an empty object. You can pass the error message and the stacktrace as a String variable (@MarkvanMents )

Author: Emir Uner
  [4 days ago](https://mendix.slack.com/archives/C046J1B7MD1/p1669295160451719?thread_ts=1669294654.900819&cid=C046J1B7MD1)
 _but with entity access disabled you can’t pass the error message to a sub-microflow_
This does not sound correct. If an object is already set to a variable, entity access does not have any affect on passing the object into sub microflows.

Paul Moes
  [4 days ago](https://mendix.slack.com/archives/C046J1B7MD1/p1669298240000329?thread_ts=1669294654.900819&cid=C046J1B7MD1)
[@MarkvM](https://mendix.slack.com/team/U8SR7G9NF)
 sorry I was wrong here, should have checked. Passing the error to a subflow as an input parameter is fine. It's just that $latestError will not be available in a subflow. Somehow this was wrong in our conventions for years :man-shrugging:

MarkvM
  [4 days ago](https://mendix.slack.com/archives/C046J1B7MD1/p1669302529554469?thread_ts=1669294654.900819&cid=C046J1B7MD1)
Thanks for the clarification. I'll put that bit back in the documentation.

{{% /alert %}}

> ## 5 Best Practices

Author: Should we put this back in? These are the activities which are mostly likely to go wrong:
Always add custom error handling on integration or email activities.
(Should we also add Java action here? (Robert Lindblom))

Paul Moes [4 days ago](https://paper.dropbox.com/doc/Error-Handling-in-Microflows-the-latest-version--BspuiTqNxxKdn9kaI4iWtYGNAg-jPf6V0YsUzdlApQmzLhU9?t=076418794601300550707256#:t=076418794601300550707256)
​​I actually find that this recommendation leads to a lot of bad implementation of error handling. Default error handling is better than a bad implementation of custom error handling. I’ve seen too many developers forgetting to apply rollback or logging the error message

Author: We added a revised best practice back to the document after technical verification: https://github.com/mendix/docs/pull/5332/commits/bfeccbf599cbf0e5b4f7dc433685b3b77f21f92a
